odoo/tests/test-uninstalls.py

```diff
@@ -0,0 +1,135 @@
+    'pad_project', 'note_pad', 'pos_cache', 'pos_blackbox_be', 'payment_test',
+}
+IGNORE = ('hw_', 'theme_', 'l10n_', 'test_', 'payment_')
+
+
+def install(db_name, module_id, module_name):
+    with odoo.api.Environment.manage():
+        with odoo.registry(db_name).cursor() as cr:
+            env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+            module = env['ir.module.module'].browse(module_id)
+            module.button_immediate_install()
+    _logger.info('%s installed', module_name)
+
+
+def uninstall(db_name, module_id, module_name):
+    with odoo.api.Environment.manage():
+        with odoo.registry(db_name).cursor() as cr:
+            env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+            module = env['ir.module.module'].browse(module_id)
+            module.button_immediate_uninstall()
+    _logger.info('%s uninstalled', module_name)
+
+
+def cycle(db_name, module_id, module_name):
+    install(db_name, module_id, module_name)
+    uninstall(db_name, module_id, module_name)
+    install(db_name, module_id, module_name)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Script for testing the install / uninstall / reinstall cycle of Odoo modules")
+    parser.add_argument("--database", "-d", type=str, required=True,
+        help="The database to test (note: must have only 'base' installed)")
+    parser.add_argument("--skip", "-s", type=str, nargs=1,
+        help="Comma-separated list of modules to skip (they will only be installed)")
+    parser.add_argument("--resume", "-r", type=str, nargs=1,
+        help="Skip the cycle on all modules up to the specified one")
+    parser.add_argument("--addons-path", "-p", type=str, nargs=1,
+        help="Comma-separated list of paths to directories containing extra Odoo modules")
+    parser.add_argument("--uninstall", "-U", type=str,
+        help="Comma-separated list of modules to uninstall/reinstall")
+    return parser.parse_args()
+
+
+def test_full(args):
+    """ Test full install/uninstall/reinstall cycle for all modules """
+    with odoo.api.Environment.manage():
+        with odoo.registry(args.database).cursor() as cr:
+            env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+
+            def valid(module):
+                return not (
+                    module.name in BLACKLIST
+                    or module.name.startswith(IGNORE)
+                    or module.state in ('installed', 'uninstallable')
+                )
+
+            modules = env['ir.module.module'].search([]).filtered(valid)
+
+            # order modules in topological order
+            modules = modules.browse(topological_sort({
+                module.id: module.dependencies_id.depend_id.ids
+                for module in modules
+            }))
+            module_names = {module.id: module.name for module in modules}
+
+    resuming = args.resume
+    for module_id, module_name in module_names.items():
```

Comment:
```suggestion
    # `module_names` preserves the order of `modules` in P3.6+
    for module_id, module_name in module_names.items():
```

Reply:
I will use a plain list instead.  The dict is not necessary anymore.